language/stdlib/modules/payment_router.move

```diff
@@ -0,0 +1,144 @@
+    acquires PaymentRouterInfo {
+        borrow_global_mut<PaymentRouterInfo>(Transaction::sender())
+            .exclusive_withdrawals_only = exclusive_withdrawals_only
+    }
+
+    // Return whether the router at `router_account_addr` only allows
+    // exlusive withdrawals
+    public fun exclusive_withdrawals_only(router_account_addr: address): bool
+    acquires PaymentRouterInfo {
+        borrow_global_mut<PaymentRouterInfo>(router_account_addr).exclusive_withdrawals_only
+    }
+
+    // Allow the account at `addr` to delegate its withdrawal_capability to us.
+    public fun allow_account_address(addr: address)
+    acquires PaymentRouterInfo {
+        let sender = Transaction::sender();
+        let router_info = borrow_global_mut<PaymentRouterInfo>(sender);
+        if (!Vector::contains(&router_info.allowed_addresses, &addr))
+            Vector::push_back(&mut router_info.allowed_addresses, addr);
+    }
+
+    // Allow routing of currencies of `Token` type.
+    public fun allow_currency<Token>() {
+        move_to_sender(AccountInfo<Token>{ child_accounts: Vector::empty() })
+    }
+
+    // Add the sending account of currency type `Token` to the router at
+    // `router_account_addr`.  The sending address must be in the
+    // `allowed_addresses` held under the router account.
+    public fun add_account_to<Token>(router_account_addr: address)
+    acquires PaymentRouterInfo, AccountInfo {
+        let sender = Transaction::sender();
+        let router_info = borrow_global_mut<PaymentRouterInfo>(router_account_addr);
+        let (has, index) = Vector::index_of(&router_info.allowed_addresses, &sender);
+        Transaction::assert(has, 0);
+        let account_info = borrow_global_mut<AccountInfo<Token>>(router_account_addr);
+        Vector::swap_remove(&mut router_info.allowed_addresses, index);
+        Vector::push_back(
+            &mut account_info.child_accounts,
+            Transaction::sender(),
+
+        );
+        move_to_sender(RoutedAccount<Token> {
+            router_account_addr,
+            withdrawal_cap: LibraAccount::extract_sender_withdrawal_capability()
+        })
+    }
+
+    // Routes deposits to a sub-account that holds currencies of type `Token`.
+    public fun deposit<Token>(router_account_addr: address, coin: Libra::T<Token>)
+    acquires AccountInfo {
+        let addrs = &borrow_global<AccountInfo<Token>>(router_account_addr).child_accounts;
+        Transaction::assert(Vector::length(addrs) > 0, 1);
+        // TODO: policy around how to rotate through different accounts
+        let index = 0;
+        LibraAccount::deposit(
+            *Vector::borrow(addrs, index),
+            coin
+        );
+    }
+
+    // Withdraws `amount` of `Token` currency from the sending account
+    // using the delegated withdrawal capability in the PaymentRouterInfo
+    // at `router_account_addr`.
+    public fun withdraw_through<Token>(router_account_addr: address, amount: u64): Libra::T<Token>
```

Comment:
Could this be rewritten without `router_account_addr`? e.g.,

```
let routed_info = borrow_global<RoutedAccount<Token>>(Transaction::sender());
let router_info = borrow_global<PaymentRouterInfo>(routed_info.router_account_addr);
Transaction::assert(!router_info.exclusive_withdrawals_only, 2);
LibraAccount::withdraw_with_capability(
  &routed_info.withdrawal_cap,
   amount
)
```

Reply:
ah! yes. I'll update :)